### PR TITLE
NUM-335 Add Next/Previous token for pagination (GetAccounts & GetTransactions)

### PR DESCRIPTION
**Pull request by antoinegelloz**

### CODE REVIEW

File: pkg/api/controllers/transaction_controller.go [hunk 1]
```diff
@@ -59,9 +63,24 @@ func (ctl *TransactionController) GetTransactions(c *gin.Context) {
...
+		}
+		t := sqlstorage.PaginationToken{}
+		if err = json.Unmarshal(res, &t); err != nil {
+			ResponseError(c, ledger.NewValidationError("invalid query value 'pagination_token'"))
+			return
+		}
+		after = strconv.FormatUint(t.ID, 10)
+	}
+
 	cursor, err := l.(*ledger.Ledger).GetTransactions(
```
Comment (gfyrag) [inline pkg/api/controllers/transaction_controller.go]: Use pagination_token with others query params (reference, account or source) can lead to unexpected behavior.
Also, client side, it is more complicated to deal with the pagination if we have to store all theses variables.
In the best of the world, those parameters are defined at the first query and never retransmitted.
So i think it should be stored inside the pagination_token and other query params should be ignored (or trigger an error) if specified.

File: pkg/storage/sqlstorage/transactions.go [hunk 2]
```diff
@@ -94,15 +94,27 @@ func (s *Store) getTransactions(ctx context.Context, exec executor, q query.Quer
 		return sharedapi.Cursor{}, s.error(err)
 	}
 
-	hasMore := false
+	previous := ""
+	if q.After != "" && len(txs) > 0 {
+		previous, err = tokenMarshal(PaginationToken{txs[0].ID + query.DefaultLimit + 1})
```
Comment (gfyrag) [inline pkg/storage/sqlstorage/transactions.go]: Are you sure?

File: pkg/api/controllers/transaction_controller.go (hunk 1 above)
Comment (antoinegelloz) [reply pkg/api/controllers/transaction_controller.go]: Ok thanks! I agree 👍

File: pkg/api/controllers/account_controller.go [hunk 3]
```diff
@@ -38,6 +41,44 @@ func (ctl *AccountController) CountAccounts(c *gin.Context) {
 func (ctl *AccountController) GetAccounts(c *gin.Context) {
 	l, _ := c.Get("ledger")
 
+	if c.Query("pagination_token") != "" {
+		if c.Query("after") != "" ||
+			c.Query("address") != "" ||
```
Comment (jdupas22) [inline pkg/api/controllers/account_controller.go:48]: if any of the filters are empty we can't make pagination ?

File: pkg/api/controllers/pagination_test.go [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+	"github.com/numary/ledger/pkg/api"
+	"github.com/numary/ledger/pkg/api/internal"
+	"github.com/numary/ledger/pkg/core"
+	"github.com/numary/ledger/pkg/ledger/query"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/fx"
+)
+
+var maxTxsPages = 3
```
Comment (jdupas22) [inline pkg/api/controllers/pagination_test.go]: maybe should be const ?

File: pkg/api/controllers/pagination_test.go [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+var maxAdditionalTxs = 2
```
Comment (jdupas22) [inline pkg/api/controllers/pagination_test.go]: same

File: pkg/api/controllers/account_controller.go (hunk 3 above)
Comment (antoinegelloz) [reply pkg/api/controllers/account_controller.go:48]: We want to ensure that if a pagination token is passed by the user, no other parameters are passed along with it.
The reason is that all the initial query params are already included in the token, so passing potentially different ones will lead to unexpected behaviour.

File: pkg/api/controllers/pagination_test.go (hunk 4 above)
Comment (antoinegelloz) [reply pkg/api/controllers/pagination_test.go]: Yes ;)

File: pkg/api/controllers/account_controller.go (hunk 3 above)
Comment (jdupas22) [reply pkg/api/controllers/account_controller.go:48]: oh alright, makes sense !

**Review by gfyrag — APPROVED** [review]

**Pull request merged**